File: native_image_cropper_android/android/src/main/kotlin/biz/cosee/native_image_cropper/NativeImageCropperPlugin.kt
package biz.cosee.native_image_cropper

import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.graphics.Canvas
import android.graphics.Matrix
import android.graphics.Paint
import android.graphics.PorterDuff
import android.graphics.PorterDuffXfermode
import android.graphics.Rect
import android.graphics.RectF
import androidx.exifinterface.media.ExifInterface
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import java.io.ByteArrayOutputStream
import java.util.concurrent.ExecutorService
import java.util.concurrent.Executors

/** The Android implementation of NativeImageCropperPlugin. */
class NativeImageCropperPlugin : FlutterPlugin, MethodCallHandler {
    companion object {
        /**
         * Creates a new cached thread pool that creates new threads as needed and reuses
         * previously created threads when they are available. This can be used to execute tasks
         * concurrently.
         */
        private val threadPool: ExecutorService = Executors.newCachedThreadPool()
    }

    /**
     * The MethodChannel that will the communication between Flutter and native Android
     * This local reference serves to register the plugin with the Flutter Engine and unregister it
     * when the Flutter Engine is detached from the Activity.
     */
    private lateinit var channel: MethodChannel

    /**
     * Initializes a [MethodChannel], which is used to communicate between the Flutter code
     * and the native Android code.
     */
    override fun onAttachedToEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        channel =
            MethodChannel(
                flutterPluginBinding.binaryMessenger,
                "biz.cosee/native_image_cropper_android"
            )
        channel.setMethodCallHandler(this)
    }

    /** Unregisters the [MethodCallHandler] from the Flutter engine. */
    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
    }

    /** Handles Flutter method calls made to this Flutter plugin. */
    override fun onMethodCall(call: MethodCall, result: Result) {
        threadPool.execute {
            when (call.method) {
                "cropRect" -> handleCropRect(call, result)
                "cropOval" -> handleCropOval(call, result)
                else -> result.notImplemented()
            }
        }
    }

    /**  Extracts the arguments from [call] and crops the image in a rectangular shape. */
    private fun handleCropRect(call: MethodCall, result: Result) {
        val bytes = call.argument<ByteArray>("bytes")!!
        val x = call.argument<Int>("x")!!
        val y = call.argument<Int>("y")!!
        val width = call.argument<Int>("width")!!
        val height = call.argument<Int>("height")!!
        val imageFormat = ImageFormat.valueOf(call.argument<String>("imageFormat")!!.uppercase())

        try {
            val croppedBitmap: Bitmap = getCroppedRectBitmap(bytes, x, y, width, height)
            result.success(bitmapToByteArray(croppedBitmap, imageFormat))
        } catch (e: IllegalArgumentException) {
            result.error("IllegalArgumentException", e.message, null)
        }
    }

    /**  Extracts the arguments from [call] and crops the image in a oval shape. */
    private fun handleCropOval(call: MethodCall, result: Result) {
        val bytes = call.argument<ByteArray>("bytes")!!
        val x = call.argument<Int>("x")!!
        val y = call.argument<Int>("y")!!
        val width = call.argument<Int>("width")!!
        val height = call.argument<Int>("height")!!
        val imageFormat = ImageFormat.valueOf(call.argument<String>("imageFormat")!!.uppercase())

        try {
            val croppedBitmap = getCroppedRectBitmap(bytes, x, y, width, height)
            val ovalBitmap = croppedBitmap.createOvalBitmap()
            result.success(bitmapToByteArray(ovalBitmap, imageFormat))
        } catch (e: IllegalArgumentException) {
            result.error("IllegalArgumentException", e.message, null)
        }
    }

    /** Converts a [Bitmap] image to a [ByteArray] by using the given [format] for compression. */
    private fun bitmapToByteArray(bitmap: Bitmap, format: ImageFormat): ByteArray {
        val stream = ByteArrayOutputStream()
        if (format == ImageFormat.JPG) {
            bitmap.compress(Bitmap.CompressFormat.JPEG, 100, stream)
        } else {
            bitmap.compress(Bitmap.CompressFormat.PNG, 100, stream)
        }
        val byteArray = stream.toByteArray()
        stream.flush()
        stream.close()
        return byteArray
    }

    /** Creates a rectangular cropped [Bitmap] of the given [bytes]. */
    private fun getCroppedRectBitmap(
        bytes: ByteArray,
        x: Int,
        y: Int,
        width: Int,
        height: Int
    ): Bitmap {
        val bitmap = BitmapFactory.decodeByteArray(bytes, 0, bytes.count())
        val exif = ExifInterface(bytes.inputStream())
        val orientation = exif.getAttributeInt(
			ExifInterface.TAG_ORIENTATION,
			ExifInterface.ORIENTATION_NORMAL
			)
	val matrix = Matrix().apply {
		setRotate(
			when(orientation) {
				ExifInterface.ORIENTATION_ROTATE_90 -> 90f
				ExifInterface.ORIENTATION_ROTATE_180 -> 180f
				ExifInterface.ORIENTATION_ROTATE_270 -> 270f
				else -> 0f
			}
		)
	}
	val rotatedBitmap = Bitmap.createBitmap(
		bitmap,
		0,
		0,
		bitmap.width,
		bitmap.height,
		matrix,
		false
	)
	val croppedBitmap = Bitmap.createBitmap(
		rotatedBitmap,
		x,
		y,
		width,
		height,
		null,
		false
	)
	bitmap.recycle()
        return croppedBitmap
    }

    /** Extension method for creating an oval [Bitmap] from a given [Bitmap]. */
    private fun Bitmap.createOvalBitmap(): Bitmap {
        val output = Bitmap.createBitmap(this.width, this.height, Bitmap.Config.ARGB_8888)
        val canvas = Canvas(output)
        val paint = Paint()
        val rect = Rect(0, 0, this.width, this.height)

        paint.isAntiAlias = true
        canvas.drawOval(
            RectF(0f, 0f, width.toFloat(), height.toFloat()),
            paint,
        )
        paint.xfermode = PorterDuffXfermode(PorterDuff.Mode.SRC_IN)
        canvas.drawBitmap(this, rect, rect, paint)

        return output
    }

    /**
     * Represents the image file format. It is used to decide how the image should be
     * compressed.
     */
    private enum class ImageFormat {
        /** Compress the image using JPG, which is usually faster. */
        JPG,

        /** Compress the image using PNG, which is lossless but slower. */
        PNG,
    }
}

File: native_image_cropper_macos/macos/Classes/NativeImageCropperPlugin.swift
import Cocoa
import FlutterMacOS

public class NativeImageCropperPlugin: NSObject, FlutterPlugin {
    public static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterMethodChannel(name: "biz.cosee/native_image_cropper_macos", binaryMessenger: registrar.messenger)
        let instance = NativeImageCropperPlugin()
        registrar.addMethodCallDelegate(instance, channel: channel)
    }
    
    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
            case "cropOval":
                handleCropOval(call, result)
            case "cropRect":
                handleCropRect(call, result)
            default:
                result(FlutterMethodNotImplemented)
        }
    }
    
    /// Extracts the arguments from [call] and crops the image in an oval shape.
    private func handleCropOval(_ call: FlutterMethodCall,_ result: @escaping FlutterResult) {
        DispatchQueue.main.async {
            do {
                let args = call.arguments as! [String: Any]
                let bytes = args["bytes"] as! FlutterStandardTypedData
                let x = args["x"] as! Int
                let y = args["y"] as! Int
                let width = args["width"] as! Int
                let height = args["height"] as! Int
                let imageFormat = ImageFormat(rawValue: (args["imageFormat"] as! String).uppercased())!
                
                let image = try self.flutterStandardTypeDataToNSImage(bytes: bytes)
                let croppedImage = try self.getCroppedRectNSImage(image: image, x: x, y: y, width: width, height: height)
                let croppedBytes = try self.nsImageToFlutterStandardTypeData(image: croppedImage, format: imageFormat)
                result(croppedBytes)
            } catch {
                result(FlutterError.init(code: String(describing: error.self), message: error.localizedDescription, details:nil))
            }
        }
    }
    
    /// Extracts the arguments from [call] and crops the image in a rectangular shape.
    private func handleCropRect(_ call: FlutterMethodCall,_ result: @escaping FlutterResult) {
        DispatchQueue.main.async {
            do {
                let args = call.arguments as! [String: Any]
                let bytes = args["bytes"] as! FlutterStandardTypedData
                let x = args["x"] as! Int
                let y = args["y"] as! Int
                let width = args["width"] as! Int
                let height = args["height"] as! Int
                let imageFormat = ImageFormat(rawValue: (args["imageFormat"] as! String).uppercased())!
                
                let image = try self.flutterStandardTypeDataToNSImage(bytes: bytes)
                let croppedImage = try self.getCroppedRectNSImage(image: image, x: x, y: y, width: width, height: height)
                let croppedBytes = try self.nsImageToFlutterStandardTypeData(image: croppedImage, format: imageFormat)
                result(croppedBytes)
            } catch {
                result(FlutterError.init(code: String(describing: error.self), message: error.localizedDescription, details:nil))
            }
        }
    }
    
    /// Creates an oval cropped [UIImage].
    private func getCroppedOvalNSImage(image: NSImage, x: Int, y: Int, width: Int, height: Int) throws -> NSImage {
       return image
    }
    
    /// Creates a rectangular cropped [NSImage].
    private func getCroppedRectNSImage(image: NSImage, x: Int, y: Int, width: Int, height: Int) throws -> NSImage {
        let cropRect = CGRect(x: x, y: y, width: width, height: height).integral
        let cgImage = image.cgImage(forProposedRect: nil, context: nil, hints: nil)
        let croppedCgImage = cgImage?.cropping(to: cropRect)
        guard let croppedCgImage  = cgImage?.cropping(to: cropRect) else {
            throw NativeImageCropperError.cgImageNullError
        }
        
        return NSImage(cgImage: croppedCgImage, size: NSSize(width: width, height: height))
    }
    
    /// Converts a [FlutterStandardTypedData] to a [NSImage].
    private func flutterStandardTypeDataToNSImage(bytes: FlutterStandardTypedData) throws -> NSImage  {
        guard let image = NSImage(data: bytes.data) else {
            throw NativeImageCropperError.flutterStandardTypeDataToNSImageError
        }
        
        return image
    }
    
    /// Converts a [NSImage] to a [FlutterStandardTypedData].
    private func nsImageToFlutterStandardTypeData(image: NSImage, format: ImageFormat) throws -> FlutterStandardTypedData {
        guard let cgImage = image.cgImage(forProposedRect: nil, context: nil, hints: nil) else {
            throw NativeImageCropperError.cgImageNullError
        }
        
        let fileType: NSBitmapImageRep.FileType
        if case .JPG = format {
            fileType = NSBitmapImageRep.FileType.jpeg
        } else {
            fileType = NSBitmapImageRep.FileType.png
        }
        
        let bitmapRep = NSBitmapImageRep(cgImage: cgImage)
        guard let bytes = bitmapRep.representation(using: fileType, properties: [:]) else {
            throw NativeImageCropperError.dataNullError
        }
        
        return FlutterStandardTypedData(bytes: bytes)
    }
    
    ///  Represents the image file format. It is used to decide how the image should be compressed.
    enum ImageFormat: String {
        /// Compress the image using JPG, which is usually faster.
        case JPG = "JPG"
        
        /// Compress the image using PNG, which is lossless but slower.
        case PNG = "PNG"
    }
}
    
private enum NativeImageCropperError: LocalizedError {
    case flutterStandardTypeDataToNSImageError
    case nsImageToFlutterStandardTypeDataError
    case cgImageNullError
    case dataNullError
}


extension NativeImageCropperError {
    public var errorDescription: String?{
        switch self {
            case .flutterStandardTypeDataToNSImageError:
                return "Could not convert FlutterStandardTypeData to NSImage."
            case .nsImageToFlutterStandardTypeDataError:
                return "Could not convert NSImage to FlutterStandartTypeData."
            case .cgImageNullError:
                return "Could not crop because CGImage is null."
            case .dataNullError:
                return "Could not crop because Data is null."
        }
    }
}
